modules/l4proxyprotocol/handler.go

```diff
@@ -47,26 +53,107 @@ func (Handler) CaddyModule() caddy.ModuleInfo {
 
 // Provision sets up the module.
 func (h *Handler) Provision(ctx caddy.Context) error {
-	if h.Timeout == 0 {
-		h.Timeout = caddy.Duration(5 * time.Second)
+	for _, s := range h.Allow {
+		_, n, err := net.ParseCIDR(s)
+		if err != nil {
+			return fmt.Errorf("invalid subnet '%s': %w", s, err)
+		}
+		h.rules = append(h.rules, proxyprotocol.Rule{Timeout: time.Duration(h.Timeout), Subnet: n})
 	}
+	h.tidyRules()
 
 	h.logger = ctx.Logger(h)
 	return nil
 }
 
+// tidyRules removes duplicate subnet rules, and use the lowest non-zero timeout.
+//
+// This is basically a copy of `Listener.SetFilter` from the proxyprotocol package.
+func (h *Handler) tidyRules() {
+	rules := h.rules
+	sort.Slice(rules, func(i, j int) bool {
+		iOnes, iBits := rules[i].Subnet.Mask.Size()
+		jOnes, jBits := rules[j].Subnet.Mask.Size()
+		if iOnes != jOnes {
+			return iOnes > jOnes
+		}
+		if iBits != jBits {
+			return iBits > jBits
+		}
+		if rules[i].Timeout != rules[j].Timeout {
+			if rules[j].Timeout == 0 {
+				return true
+			}
+			return rules[i].Timeout < rules[j].Timeout
+		}
+		return rules[i].Timeout < rules[j].Timeout
+	})
+
+	if len(rules) > 0 {
+		// dedup
+		last := rules[0]
+		nf := rules[1:1]
+		for _, f := range rules[1:] {
+			if last.Subnet.String() == f.Subnet.String() {
+				continue
+			}
+
+			last = f
+			nf = append(nf, f)
+		}
+	}
+}
+
+// newConn creates a new connection which will handle the PROXY protocol. It
+// will return nil if the remote IP does not match the allowable CIDR ranges.
+//
+// This is basically a copy of `Listener.Accept` from the proxyprotocol package.
+func (h *Handler) newConn(cx *layer4.Connection) *proxyprotocol.Conn {
+	nc := func(t time.Duration) *proxyprotocol.Conn {
+		if t == 0 {
+			return proxyprotocol.NewConn(cx, time.Time{})
+		}
+		return proxyprotocol.NewConn(cx, time.Now().Add(t))
+	}
+
+	if len(h.rules) == 0 {
+		return nc(time.Duration(h.Timeout))
+	}
+
+	var remoteIP net.IP
+	switch r := cx.RemoteAddr().(type) {
+	case *net.TCPAddr:
+		remoteIP = r.IP
+	case *net.UDPAddr:
+		remoteIP = r.IP
+	default:
+		return nil
+	}
+
+	for _, r := range h.rules {
+		if r.Subnet.Contains(remoteIP) {
+			return nc(r.Timeout)
+		}
+	}
+
+	return nil
+}
+
 // Handle handles the connections.
 func (h *Handler) Handle(cx *layer4.Connection, next layer4.Handler) error {
-	var deadline time.Time
-	if h.Timeout != 0 {
-		deadline = time.Now().Add(time.Duration(h.Timeout))
+	conn := h.newConn(cx)
+	if conn == nil {
+		h.logger.Error("remote addr not allowed",
+			zap.String("remote", cx.RemoteAddr().String()),
+			zap.Strings("allow", h.Allow),
+		)
+		return next.Handle(cx)
 	}
```

Comment:
It might be prohibitively noisy to emit `Error` level logs if some bad actor makes PROXY requests from a rejected IP. For certain errors like this in Caddy, we tend to use `Debug` level so that a bad actor can't just cause the logs to fill up fast.

Reply:
Couldn't agree more!